test/run-drun/stable-log.mo

```diff
@@ -0,0 +1,73 @@
+import Prim "mo:⛔";
+import StableMemory "stable-mem/StableMemory";
+
+actor {
+
+  func ensure(offset : Nat32) {
+    let pages = (offset + 65536) >> 16;
+      if (pages > StableMemory.size()) {
+        let oldsize = StableMemory.grow(pages - StableMemory.size());
+        assert (oldsize != 0xFFFF);
+     };
+  };
+
+  stable var base : Nat32 = 0;
+
+  public func log(t : Text) {
+    let blob = Prim.encodeUtf8(t);
+    let size = Prim.natToNat32(blob.size());
+    let next = base + 4 + size;
+    ensure(next);
+    StableMemory.storeNat32(base, size);
+    StableMemory.storeBlob(base + 4, blob);
+    base := next;
+  };
+
+  public query func read(offset : Nat32) : async ?(Text, nextOffset : Nat32) {
+    if (offset >= base) return null;
+    let size = StableMemory.loadNat32(offset);
+    let blob = StableMemory.loadBlob(offset + 4, Prim.nat32ToNat(size));
+    let ?text = Prim.decodeUtf8(blob);
+    ?(text, offset + 4 + size)
```

Comment:
Would be nice to avoid the warning. I think you can use `do?`:
```suggestion
    do? { (Prim.decodeUtf8(blob)!, offset + 4 + size) }
```

Reply:
Done